crates/bevy_time/src/virt.rs

```diff
@@ -0,0 +1,423 @@
+/// laptop is suspended for an hour, it doesn't make sense to try to simulate
+/// the game logic for the elapsed hour when resuming. Instead it is better to
+/// lose the extra time and pretend a shorter duration of time passed. Setting
+/// [`max_delta()`](Time::max_delta) to a relatively short time means that the
+/// impact on game logic will be minimal.
+///
+/// If the game lags for some reason, meaning that it will take a longer time to
+/// compute a frame than the real time that passes during the computation, then
+/// we would fall behind in processing virtual time. If this situation persists,
+/// and computing a frame takes longer depending on how much virtual time has
+/// passed, the game would enter a "death spiral" where computing each frame
+/// takes longer and longer and the game will appear to freeze. By limiting the
+/// maximum time that can be added at once, we also limit the amount of virtual
+/// time the game needs to compute for each frame. This means that the game will
+/// run slow, and it will run slower than real time, but it will not freeze and
+/// it will recover as soon as computation becomes fast again.
+///
+/// You should set [`max_delta()`](Time::max_delta) to a value that is
+/// approximately the minimum FPS your game should have even if heavily lagged
+/// for a moment. The actual FPS when lagged will be somewhat lower than this,
+/// depending on how much more time it takes to compute a frame compared to real
+/// time. You should also consider how stable your FPS is, as the limit will
+/// also dictate how big of an FPS drop you can accept without losing time and
+/// falling behind real time.
+#[derive(Debug, Copy, Clone, Reflect)]
+pub struct Virtual {
+    max_delta: Duration,
+    paused: bool,
+    relative_speed: f64,
+    effective_speed: f64,
+}
+
+impl Time<Virtual> {
+    const DEFAULT_MAX_DELTA: Duration = Duration::from_millis(333); // XXX: better value
+
+    /// Create new virtual clock with given maximum delta step
+    /// [`Duration`](std::time::Duration)
+    pub fn from_max_delta(max_delta: Duration) -> Self {
+        let mut ret = Self::default();
+        ret.set_max_delta(max_delta);
+        ret
+    }
+
+    /// Returns the maximum amount of time that can be added to this clock by a
+    /// single update, as [`Duration`](std::time::Duration).
+    ///
+    /// This is the maximum value [`Self::delta()`] will return and also to
+    /// maximum time [`Self::elapsed()`] will be increased by in a single
+    /// update.
+    ///
+    /// This ensures that even if there are no updates scheduled for an extended
+    /// amount of time, there won't be a huge update at once for the clock. This
+    /// also limits the number of fixed time ticks that will be evaluated at
+    /// maximum per a single update.
+    ///
+    /// The default value is 0.333 seconds **FIXME**.
+    #[inline]
+    pub fn max_delta(&self) -> Duration {
```

Comment:
```suggestion
    pub const fn max_delta(&self) -> Duration {
```

Reply:
And so on throughtout this PR. It's nice to make methods const where possible, especially when we never expect to need to break that guarantee.

Reply:
Decided against this, discussion was in Discord